network/src/config.rs

```diff
@@ -50,56 +57,71 @@ impl Config {
+        sync_provider_addresses: Vec<String>,
         peer_sync_interval: Duration,
     ) -> Result<Self, NetworkError> {
-        // Convert the given bootnodes into socket addresses.
-        let mut bootnodes = Vec::with_capacity(bootnodes_addresses.len());
-        for bootnode_address in bootnodes_addresses.iter() {
-            if let Ok(bootnode) = bootnode_address.parse::<SocketAddr>() {
-                bootnodes.push(bootnode);
-            }
-        }
+        // Convert the given seeded nodes into socket addresses.
+        let beacons: Vec<SocketAddr> = beacon_addresses
+            .into_iter()
+            .flat_map(|addr| addr.parse().ok())
+            .collect();
+
+        let sync_providers: Vec<SocketAddr> = sync_provider_addresses
+            .into_iter()
+            .filter_map(|addr| addr.parse().ok())
+            .collect();
 
         Ok(Self {
             node_id,
+            node_type,
             desired_address,
             minimum_number_of_connected_peers,
             maximum_number_of_connected_peers,
-            bootnodes: ArcSwap::new(Arc::new(bootnodes)),
-            is_bootnode,
-            is_crawler,
+            beacons: ArcSwap::new(Arc::new(beacons)),
+            sync_providers: ArcSwap::new(Arc::new(sync_providers)),
             peer_sync_interval,
         })
     }
 
-    /// Returns the default bootnodes of the network.
+    /// Returns the default peer discovery nodes of the network.
     #[inline]
-    pub fn bootnodes(&self) -> Arc<Vec<SocketAddr>> {
-        self.bootnodes.load_full()
+    pub fn beacons(&self) -> Arc<Vec<SocketAddr>> {
+        self.beacons.load_full()
     }
 
-    /// Returns `true` if this node is a bootnode. Otherwise, returns `false`.
+    /// Returns the default sync provider nodes of the network.
     #[inline]
-    pub fn is_bootnode(&self) -> bool {
-        self.is_bootnode
+    pub fn sync_providers(&self) -> Arc<Vec<SocketAddr>> {
+        self.sync_providers.load_full()
+    }
+
+    /// Returns `true` if this node is a sync provider node. Otherwise, returns `false`.
+    #[inline]
+    pub fn is_sync_provider(&self) -> bool {
+        matches!(self.node_type, NodeType::SyncProvider)
     }
 
     /// Returns `true` if this node is a crawler. Otherwise, returns `false`.
     #[inline]
     pub fn is_crawler(&self) -> bool {
-        self.is_crawler
+        matches!(self.node_type, NodeType::Crawler)
+    }
+
+    /// Returns `true` if this node is a plain client node. Otherwise, returns `false`.
+    #[inline]
+    pub fn is_client(&self) -> bool {
+        matches!(self.node_type, NodeType::Client)
     }
 
-    /// Returns `true` if this node is a plain node. Otherwise, returns `false`.
+    /// Returns `true` if this node is a peer discovery node. Otherwise, returns `false`.
     #[inline]
-    pub fn is_regular_node(&self) -> bool {
-        !(self.is_bootnode() || self.is_crawler())
+    pub fn is_beacon(&self) -> bool {
```

Comment:
I'm wondering if we couldn't deduplicate these methods into a single `is_of_type(t: NodeType)` or something like that

Reply:
Might need a macro for this, `matches!(self.node_type, t)` [won't actually work](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=5d8d78bc9f84119df61805a80945633a). I've been refactoring the network crate on another branch, I'll make a note to explore this change there?

Reply:
Actually, we've got PartialEq 👀